File: 15-02-view-group/java-15-02-view-group/app/src/main/java/com/example/viewgroup/EditViewGroup.java
package com.example.viewgroup;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

public class EditViewGroup extends LinearLayout {

    public EditViewGroup(Context context) {
        super(context);
        init();
    }

    public EditViewGroup(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        setOrientation(LinearLayout.VERTICAL);
        final Context context = getContext();
        final LayoutParams layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        final TextView textView = new TextView(context);
        textView.setText("Hello EditViewGroup!");
        addView(textView, layoutParams);
        final ToastEditText toastEditText = new ToastEditText(context);
        addView(toastEditText, layoutParams);
        toastEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                textView.setText("Length is: " + toastEditText.getText().length());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }
}

File: 15-02-view-group/kotlin-15-02-view-group/app/src/main/java/com/example/viewgroup/EditViewGroup.kt
package com.example.viewgroup

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.widget.LinearLayout
import android.widget.TextView

class EditViewGroup : LinearLayout {

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    private fun init() {
        orientation = LinearLayout.VERTICAL
        val context = context
        val layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
        val textView = TextView(context)
        textView.text = "Hello EditViewGroup!"
        addView(textView, layoutParams)
        val toastEditText = ToastEditText(context)
        addView(toastEditText, layoutParams)
        toastEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                textView.text = "Length is: " + toastEditText.text!!.length
            }

            override fun afterTextChanged(s: Editable) {

            }
        })
    }
}
